### ISSUE TITLE
Error in cluster_network with snakemake and thread.py

### ISSUE THREAD

**Issue by MichelJD** [PyPSA/pypsa-eur#221]
## Checklist

- [ x] I am using the current [`master`](https://github.com/PyPSA/pypsa-eur/tree/master) branch or the latest [release](https://github.com/PyPSA/pypsa-eur/releases). Please indicate.
- [ x] I am running on an up-to-date [`pypsa-eur` environment](https://github.com/PyPSA/pypsa-eur/blob/master/envs/environment.yaml). Update via `conda env update -f envs/environment.yaml`.

## Describe the Bug
Hi together, 

I an running into an error, when trying to reproduce the tutorial in the cluster_network rule. A lot of jobs are already done, this is why the job list is already quite short. I did research on the error, but couldn't find any help. So any help is highly appreciated :)

Thanks a lot in advance!

## Error Message

```
(pypsa-eur) Michels-MBP:pypsa-eur Michel$ snakemake -j1 results/networks/elec_s_6_ec_lcopt_Co2L-24H.nc
Restricted license - for non-production use only - expires 2022-01-13
No parameters matching '_test' found
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	add_extra_components
	1	cluster_network
	1	prepare_network
	1	solve_network
	4
Select jobs to execute...

[Wed Feb 24 14:22:40 2021]
rule cluster_network:
    input: networks/elec_s.nc, resources/regions_onshore_elec_s.geojson, resources/regions_offshore_elec_s.geojson, resources/busmap_elec_s.csv, data/costs.csv
    output: networks/elec_s_6.nc, resources/regions_onshore_elec_s_6.geojson, resources/regions_offshore_elec_s_6.geojson, resources/busmap_elec_s_6.csv, resources/linemap_elec_s_6.csv
    log: logs/cluster_network/elec_s_6.log
    jobid: 3
    benchmark: benchmarks/cluster_network/elec_s_6
    wildcards: simpl=, clusters=6
    resources: mem=3000

/opt/anaconda3/envs/pypsa-eur/lib/python3.8/site-packages/pypsa/io.py:771: FutureWarning: Index.__or__ operating as a set operation is deprecated, in the future this will be a logical operation matching Series.__or__.  Use index.union(other) instead
  pnl[attr] = pnl[attr].reindex(columns=(pnl[attr].columns | columns))
INFO:pypsa.io:Imported network elec_s.nc has buses, carriers, generators, lines, links, loads, storage_units
Traceback (most recent call last):
  File "/Users/Michel/Documents/Programming/projects/pypsa-eur/.snakemake/scripts/tmpi7lseb_2.cluster_network.py", line 379, in <module>
    clustering = clustering_for_n_clusters(n, n_clusters, custom_busmap, aggregate_carriers,
  File "/Users/Michel/Documents/Programming/projects/pypsa-eur/.snakemake/scripts/tmpi7lseb_2.cluster_network.py", line 280, in clustering_for_n_clusters
    busmap = busmap_for_n_clusters(n, n_clusters, solver_name, focus_weights, algorithm)
  File "/Users/Michel/Documents/Programming/projects/pypsa-eur/.snakemake/scripts/tmpi7lseb_2.cluster_network.py", line 236, in busmap_for_n_clusters
    n_clusters = distribute_clusters(n, n_clusters, focus_weights=focus_weights, solver_name=solver_name)
  File "/Users/Michel/Documents/Programming/projects/pypsa-eur/.snakemake/scripts/tmpi7lseb_2.cluster_network.py", line 190, in distribute_clusters
    assert n_clusters >= len(N) and n_clusters <= N.sum(), \
AssertionError: Number of clusters must be 37 <= n_clusters <= 3476 for this selection of countries.
[Wed Feb 24 14:23:02 2021]
Error in rule cluster_network:
    jobid: 3
    output: networks/elec_s_6.nc, resources/regions_onshore_elec_s_6.geojson, resources/regions_offshore_elec_s_6.geojson, resources/busmap_elec_s_6.csv, resources/linemap_elec_s_6.csv
    log: logs/cluster_network/elec_s_6.log (check log file(s) for error message)

RuleException:
CalledProcessError in line 265 of /Users/Michel/Documents/Programming/projects/pypsa-eur/Snakefile:
Command 'set -euo pipefail;  /opt/anaconda3/envs/pypsa-eur/bin/python3.8 /Users/Michel/Documents/Programming/projects/pypsa-eur/.snakemake/scripts/tmpi7lseb_2.cluster_network.py' returned non-zero exit status 1.
  File "/opt/anaconda3/envs/pypsa-eur/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2272, in run_wrapper
  File "/Users/Michel/Documents/Programming/projects/pypsa-eur/Snakefile", line 265, in __rule_cluster_network
  File "/opt/anaconda3/envs/pypsa-eur/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 568, in _callback
  File "/opt/anaconda3/envs/pypsa-eur/lib/python3.8/concurrent/futures/thread.py", line 57, in run
  File "/opt/anaconda3/envs/pypsa-eur/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 554, in cached_or_run
  File "/opt/anaconda3/envs/pypsa-eur/lib/python3.8/site-packages/snakemake/executors/__init__.py", line 2357, in run_wrapper
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /Users/Michel/Documents/Programming/projects/pypsa-eur/.snakemake/log/2021-02-24T142239.741350.snakemake.log
```

I also attached the installed package list, but it should be the current environment, as I just did the installation:

```
(pypsa-eur) Michels-MBP:pypsa-eur Michel$ conda list
# packages in environment at /opt/anaconda3/envs/pypsa-eur:
#
# Name                    Version                   Build  Channel
affine                    2.3.0                      py_0    conda-forge
ampl-mp                   3.1.0             h1423387_1004    conda-forge
amply                     0.1.4                      py_0    conda-forge
appdirs                   1.4.4              pyh9f0ad1d_0    conda-forge
appnope                   0.1.2            py38h50d1736_1    conda-forge
atlite                    0.0.3                      py_0    conda-forge
attrs                     20.3.0             pyhd3deb0d_0    conda-forge
backcall                  0.2.0              pyh9f0ad1d_0    conda-forge
backports                 1.0                        py_2    conda-forge
backports.functools_lru_cache 1.6.1                      py_0    conda-forge
beautifulsoup4            4.9.3              pyhb0f4dca_0    conda-forge
blosc                     1.21.0               h046ec9c_0    conda-forge
bokeh                     2.2.3            py38h50d1736_0    conda-forge
boost-cpp                 1.72.0               hdf9ef73_0    conda-forge
bottleneck                1.3.2            py38hc7193ba_3    conda-forge
brotlipy                  0.7.0           py38h5406a74_1001    conda-forge
bzip2                     1.0.8                hc929b4f_4    conda-forge
c-ares                    1.17.1               hc929b4f_0    conda-forge
ca-certificates           2020.12.5            h033912b_0    conda-forge
cairo                     1.16.0            hec6a9b0_1003    conda-forge
cartopy                   0.17.0          py38h9bcff04_1015    conda-forge
cdsapi                    0.5.1                    pypi_0    pypi
certifi                   2020.12.5        py38h50d1736_1    conda-forge
cffi                      1.14.5           py38ha97d567_0    conda-forge
cfitsio                   3.470                h01dc385_7    conda-forge
cftime                    1.4.1            py38ha1b04c9_0    conda-forge
chardet                   4.0.0            py38h50d1736_1    conda-forge
click                     7.1.2              pyh9f0ad1d_0    conda-forge
click-plugins             1.1.1                      py_0    conda-forge
cligj                     0.7.1              pyhd8ed1ab_0    conda-forge
cloudpickle               1.6.0                      py_0    conda-forge
coincbc                   2.10.5               he509fd4_1    conda-forge
conda                     4.9.2            py38h50d1736_0    conda-forge
conda-package-handling    1.7.2            py38h94c058a_0    conda-forge
configargparse            1.3                pyhd8ed1ab_0    conda-forge
countrycode               0.2                      pypi_0    pypi
cryptography              3.4.4            py38h43df06b_0    conda-forge
curl                      7.71.1               hcb81553_8    conda-forge
cycler                    0.10.0                     py_2    conda-forge
cytoolz                   0.11.0           py38h5406a74_3    conda-forge
dask                      2021.2.0           pyhd8ed1ab_0    conda-forge
dask-core                 2021.2.0           pyhd8ed1ab_0    conda-forge
datrie                    0.8.2            py38h4d0b108_1    conda-forge
decorator                 4.4.2                      py_0    conda-forge
descartes                 1.1.0                      py_4    conda-forge
distributed               2021.2.0         py38h50d1736_0    conda-forge
docutils                  0.16             py38h50d1736_3    conda-forge
entsoe-py                 0.2.10                     py_0    conda-forge
et_xmlfile                1.0.1                   py_1001    conda-forge
expat                     2.2.10               h1c7c35f_0    conda-forge
fiona                     1.8.13           py38he71f6a4_0    conda-forge
fontconfig                2.13.1            hd23ceaa_1004    conda-forge
freetype                  2.10.4               h4cff582_1    conda-forge
freexl                    1.0.5             h0b31af3_1002    conda-forge
fsspec                    0.8.5              pyhd8ed1ab_0    conda-forge
gdal                      3.0.4           py38h9f7df5a_10    conda-forge
geographiclib             1.50                       py_0    conda-forge
geokit                    1.1.2                    pypi_0    pypi
geopandas                 0.8.2              pyhd8ed1ab_0    conda-forge
geopy                     2.1.0              pyhd3deb0d_0    conda-forge
geos                      3.8.1                h4a8c4bd_0    conda-forge
geotiff                   1.6.0                hd8796ba_0    conda-forge
gettext                   0.19.8.1          h7937167_1005    conda-forge
giflib                    5.2.1                hbcb3906_2    conda-forge
gitdb                     4.0.5              pyhd8ed1ab_1    conda-forge
gitpython                 3.1.13             pyhd8ed1ab_0    conda-forge
glaes                     1.1.2                    pypi_0    pypi
glib                      2.66.7               h1c7c35f_0    conda-forge
glib-tools                2.66.7               h1c7c35f_0    conda-forge
gurobi                    9.1.1                    py38_0    gurobi
hdf4                      4.2.13            h71d84a9_1004    conda-forge
hdf5                      1.10.6          nompi_hc5d9132_1114    conda-forge
heapdict                  1.0.1                      py_0    conda-forge
icu                       64.2                 h6de7cb9_1    conda-forge
idna                      2.10               pyh9f0ad1d_0    conda-forge
importlib-metadata        3.4.0            py38h50d1736_0    conda-forge
importlib_metadata        3.4.0                hd8ed1ab_0    conda-forge
ipopt                     3.13.4               h8d9dcd1_0    conda-forge
ipython                   7.20.0           py38h9bb44b7_2    conda-forge
ipython_genutils          0.2.0                      py_1    conda-forge
jdcal                     1.4.1                      py_0    conda-forge
jedi                      0.18.0           py38h50d1736_2    conda-forge
jinja2                    2.11.3             pyh44b312d_0    conda-forge
joblib                    1.0.1              pyhd8ed1ab_0    conda-forge
jpeg                      9d                   hbcb3906_0    conda-forge
json-c                    0.13.1            h575e443_1002    conda-forge
jsonschema                3.2.0                      py_2    conda-forge
jupyter_core              4.7.1            py38h50d1736_0    conda-forge
kealib                    1.4.14               ha0ce315_1    conda-forge
kiwisolver                1.3.1            py38hd9c93a9_1    conda-forge
krb5                      1.17.2               h60d9502_0    conda-forge
lcms2                     2.12                 h577c468_0    conda-forge
libarchive                3.5.1                h078526d_0    conda-forge
libblas                   3.9.0                8_openblas    conda-forge
libcblas                  3.9.0                8_openblas    conda-forge
libcurl                   7.71.1               h9bf37e3_8    conda-forge
libcxx                    11.0.1               habf9029_0    conda-forge
libdap4                   3.20.6               h75eb3ff_1    conda-forge
libedit                   3.1.20191231         h0678c8f_2    conda-forge
libev                     4.33                 haf1e3a3_1    conda-forge
libffi                    3.3                  h046ec9c_2    conda-forge
libgdal                   3.0.4               h242383b_10    conda-forge
libgfortran               5.0.0           9_3_0_h6c81a4c_18    conda-forge
libgfortran5              9.3.0               h6c81a4c_18    conda-forge
libglib                   2.66.7               h56565ab_0    conda-forge
libiconv                  1.16                 haf1e3a3_0    conda-forge
libkml                    1.3.0             h88bc94a_1012    conda-forge
liblapack                 3.9.0                8_openblas    conda-forge
libnetcdf                 4.7.4           nompi_h9d8a93f_107    conda-forge
libnghttp2                1.43.0               h07e645a_0    conda-forge
libopenblas               0.3.12          openmp_h54245bb_1    conda-forge
libpng                    1.6.37               h7cec526_2    conda-forge
libpq                     12.3                 h7049927_3    conda-forge
libsolv                   0.7.17               hfd3ada9_0    conda-forge
libspatialindex           1.9.3                h1c7c35f_3    conda-forge
libspatialite             4.3.0a            h658e6c1_1038    conda-forge
libssh2                   1.9.0                h8a08a2b_5    conda-forge
libtiff                   4.2.0                h355d032_0    conda-forge
libwebp-base              1.2.0                hbcf498f_0    conda-forge
libxml2                   2.9.10               h53d96d6_0    conda-forge
libxslt                   1.1.33               h320ff13_0    conda-forge
llvm-openmp               11.0.1               h7c73e74_0    conda-forge
locket                    0.2.0                      py_2    conda-forge
lxml                      4.6.2            py38h7d97eb6_1    conda-forge
lz4-c                     1.9.2                hb1e8313_3    conda-forge
lzo                       2.10              haf1e3a3_1000    conda-forge
mamba                     0.7.14           py38hb989c8f_0    conda-forge
markupsafe                1.1.1            py38h5406a74_3    conda-forge
matplotlib-base           3.3.4            py38hb24f337_0    conda-forge
memory_profiler           0.58.0                     py_0    conda-forge
metis                     5.1.0             h2e338ed_1006    conda-forge
mock                      4.0.3            py38h50d1736_1    conda-forge
msgpack-python            1.0.2            py38hd9c93a9_1    conda-forge
mumps-include             5.2.1               h694c41f_10    conda-forge
mumps-seq                 5.2.1               h994990c_10    conda-forge
munch                     2.5.0                      py_0    conda-forge
nbformat                  5.1.2              pyhd8ed1ab_1    conda-forge
ncurses                   6.2                  h2e338ed_4    conda-forge
netcdf4                   1.5.6           nompi_py38h1f593c7_100    conda-forge
networkx                  2.5                        py_0    conda-forge
nose                      1.3.7                   py_1006    conda-forge
numexpr                   2.7.2            py38he9f00de_0    conda-forge
numpy                     1.19.0           py38h598c1e0_0    conda-forge
olefile                   0.46               pyh9f0ad1d_1    conda-forge
openjpeg                  2.3.1                h6cbf5cd_3    conda-forge
openpyxl                  3.0.6              pyhd8ed1ab_0    conda-forge
openssl                   1.1.1j               hbcf498f_0    conda-forge
owslib                    0.23.0             pyhd8ed1ab_0    conda-forge
packaging                 20.9               pyh44b312d_0    conda-forge
pandas                    1.2.2            py38hb77cc89_0    conda-forge
parso                     0.8.1              pyhd8ed1ab_0    conda-forge
partd                     1.1.0                      py_0    conda-forge
patsy                     0.5.1                      py_0    conda-forge
pcre                      8.44                 hb1e8313_0    conda-forge
pexpect                   4.8.0              pyh9f0ad1d_2    conda-forge
pickleshare               0.7.5                   py_1003    conda-forge
pillow                    8.1.0            py38h4c06724_2    conda-forge
pip                       21.0.1             pyhd8ed1ab_0    conda-forge
pixman                    0.38.0            h01d97ff_1003    conda-forge
ply                       3.11                       py_1    conda-forge
poppler                   0.87.0               h3232a60_1    conda-forge
poppler-data              0.4.10                        0    conda-forge
postgresql                12.3                 he852019_3    conda-forge
powerplantmatching        0.4.8              pyhd8ed1ab_0    conda-forge
progressbar2              3.53.1             pyh9f0ad1d_0    conda-forge
proj                      7.0.0                h45baca5_5    conda-forge
prompt-toolkit            3.0.16             pyha770c72_0    conda-forge
psutil                    5.8.0            py38h5406a74_1    conda-forge
ptyprocess                0.7.0              pyhd3deb0d_0    conda-forge
pulp                      2.3.1            py38h32f6830_0    conda-forge
pycosat                   0.6.3           py38h5406a74_1006    conda-forge
pycountry                 20.7.3             pyh9f0ad1d_0    conda-forge
pycparser                 2.20               pyh9f0ad1d_2    conda-forge
pyepsg                    0.4.0                      py_0    conda-forge
pygments                  2.8.0              pyhd8ed1ab_0    conda-forge
pykdtree                  1.3.4            py38hf3d72b9_0    conda-forge
pyomo                     5.7.3            py38h91a8764_0    conda-forge
pyopenssl                 20.0.1             pyhd8ed1ab_0    conda-forge
pyparsing                 2.4.7              pyh9f0ad1d_0    conda-forge
pyproj                    2.6.1.post1      py38h7d7e87d_0    conda-forge
pypsa                     0.17.1             pyh9f0ad1d_0    conda-forge
pyrsistent                0.17.3           py38h5406a74_2    conda-forge
pyshp                     2.1.3              pyh44b312d_0    conda-forge
pysocks                   1.7.1            py38h50d1736_3    conda-forge
pytables                  3.6.1            py38hfb086ad_3    conda-forge
python                    3.8.8           h4e93d89_0_cpython    conda-forge
python-dateutil           2.8.1                      py_0    conda-forge
python-utils              2.5.5              pyh44b312d_0    conda-forge
python_abi                3.8                      1_cp38    conda-forge
pytz                      2021.1             pyhd8ed1ab_0    conda-forge
pyutilib                  6.0.0              pyh9f0ad1d_0    conda-forge
pyyaml                    5.4.1            py38h5406a74_0    conda-forge
rasterio                  1.1.5            py38hac06332_1    conda-forge
ratelimiter               1.2.0                   py_1002    conda-forge
readline                  8.0                  h0678c8f_2    conda-forge
reproc                    14.2.1               hbcb3906_0    conda-forge
reproc-cpp                14.2.1               h2e338ed_0    conda-forge
requests                  2.25.1             pyhd3deb0d_0    conda-forge
rtree                     0.9.7            py38hc59ffc2_1    conda-forge
ruamel_yaml               0.15.80         py38hca655e8_1004    conda-forge
scikit-learn              0.24.1           py38hfd19401_0    conda-forge
scipy                     1.6.0            py38ha68b950_0    conda-forge
scotch                    6.0.9                h57311db_1    conda-forge
seaborn                   0.11.1               hd8ed1ab_1    conda-forge
seaborn-base              0.11.1             pyhd8ed1ab_1    conda-forge
setuptools                49.6.0           py38h50d1736_3    conda-forge
shapely                   1.7.1            py38h7843d1f_1    conda-forge
six                       1.15.0             pyh9f0ad1d_0    conda-forge
sklearn                   0.0                      pypi_0    pypi
smmap                     3.0.5              pyh44b312d_0    conda-forge
snakemake-minimal         5.32.2                     py_0    bioconda
snuggs                    1.4.7                      py_0    conda-forge
sortedcontainers          2.3.0              pyhd8ed1ab_0    conda-forge
soupsieve                 2.0.1                      py_1    conda-forge
sqlite                    3.34.0               h17101e1_0    conda-forge
statsmodels               0.12.2           py38ha1b04c9_0    conda-forge
tbb                       2020.2               h926bf3e_3    conda-forge
tblib                     1.6.0                      py_0    conda-forge
threadpoolctl             2.1.0              pyh5ca1d4c_0    conda-forge
tiledb                    1.7.7                h84aa2a7_3    conda-forge
tk                        8.6.10               h0419947_1    conda-forge
toolz                     0.11.1                     py_0    conda-forge
toposort                  1.6                pyhd8ed1ab_0    conda-forge
tornado                   6.1              py38h5406a74_1    conda-forge
tqdm                      4.57.0             pyhd8ed1ab_0    conda-forge
traitlets                 5.0.5                      py_0    conda-forge
tsam                      1.1.0                    pypi_0    pypi
typing_extensions         3.7.4.3                    py_0    conda-forge
tzcode                    2021a                h35c211d_0    conda-forge
urllib3                   1.26.3             pyhd8ed1ab_0    conda-forge
vresutils                 0.3.1                    pypi_0    pypi
wcwidth                   0.2.5              pyh9f0ad1d_2    conda-forge
wheel                     0.36.2             pyhd3deb0d_0    conda-forge
wrapt                     1.12.1           py38h5406a74_3    conda-forge
xarray                    0.16.2             pyhd8ed1ab_0    conda-forge
xerces-c                  3.2.2             h8f8adb3_1004    conda-forge
xlrd                      2.0.1              pyhd8ed1ab_3    conda-forge
xz                        5.2.5                haf1e3a3_1    conda-forge
yaml                      0.2.5                haf1e3a3_0    conda-forge
zict                      2.0.0                      py_0    conda-forge
zipp                      3.4.0                      py_0    conda-forge
zlib                      1.2.11            h7795811_1010    conda-forge
zstd                      1.4.8                h62961b2_0    conda-forge
```

**Comment by fneum** [comment]
Hiya @MichelJD,

you can find the problem to the solution in this line

> ```AssertionError: Number of clusters must be 37 <= n_clusters <= 3476 for this selection of countries.```

The default `config.default.yaml` encompasses most countries of the ENTSO-E area and it is the requirement of the clustering that each country/synchronous zone is represented by at least 1 node --> minimum 37 clusters

To reproduce the tutorial, you should use `config.tutorial.yaml`, which covers only Germany by default, here you can go down to as little as one cluster.

HTH

**Comment by MichelJD** [comment]
Hi @fneum 

thanks for your quick reply! I was thinking the same, when I went through the message, but I made sure, that I copied the config.tutorial.yaml to config.yaml. So in my config.yaml. file there is only Germany selected as a country and the number of clusters is 5. In an earlier session I ran the default config.yaml file before with all countries and more . So the program still seems to process with the old options. Anyway, I made a new clone and used the tutorial options from the beginning and it worked just fine. Is that how it is supposed to be? I thought the idea was being able to run different scenarios. What do I have to do, when changing configuration input?

Sorry, I am still in the beginning to work with pypsa-eur, but I am really looking forward of using it!

Cheers and have a good day
MJD
